### PR TITLE
Log Console extension to capture unhandled messages and other activity

### LINKED ISSUE #1254: Output from notebook comm messages
While investigating https://github.com/ipython/ipywidgets/issues/44, we uncovered an interesting situation in JLab: now that widgets *are* outputs, it's not clear anymore where output should go that is generated in response to a widget comm message. In the classic notebook, we had a hack where we would find the cell associated with a view, and put the output in that cell. But sometimes the widget change is not done in a specific view (as noted in https://github.com/ipython/ipywidgets/issues/316). And in JLab, there's not a clear place to put extra output generated from another output.

In the special case of a `stdin` message from `input()`, an output can act as an input, but the input control is immediately replaced with the value.

In the case of a widget, where an output is acting as an input that needs to stay active, it's not clear where output should go. @fperez and I talked a lot about this today, and came up with the following:

* If it's important to have the output go on the page, then capture the output explicitly in an output widget and display the output widget. This is what interact does in JLab, for example.
* Otherwise, the output messages will go to the notebook and won't be claimed. 

Here is the TODO:

- [ ] Have a notebook-level log widget that can be opened from a command and maybe a toolbar button.
- [ ] Save unclaimed output messages for the most recent request (based on the parent header)
- [ ] Pop up a notification in the corner that there was output not shown on the page available in the log. 
- [ ] When the user clicks on a link in the notification, show the log widget in a new window with a log of rendered output, and append all unclaimed output to the log window. The log should separate output from different requests, perhaps with a horizontal line. 
- [ ] Notifications will no longer show up in the notebook as long as an output log is open. 
- [ ] Notifications also have an option to dismiss or dismiss all such notifications. 
- [ ] The output log also has instructions on how to capture output using the output widget. 
- [ ] Also, there might be an easy way to drag log outputs into a notebook.
- [ ] The log should be cleared when restarting the notebook? Or perhaps just when 'Clear Output' is run? Perhaps there should be a button in the log to clear it.
- [ ] Should unclaimed messages be persisted in the notebook?

### PR DESCRIPTION
**Pull request by mbektasbbg**
OutputConsole

- Output Console intercepts all kernel messages from all kernel sessions and displays in a new panel
- IO messages of type ['execute_result', 'display_data', 'stream', 'error', 'update_display_data'] `received` from kernel are displayed
- OutputArea component is used to render the messages
- Once a new kernel message arrives a notification is shown on status bar. Status item shows number of messages and flashes one time once a message is received. If Output Console is open, it doesn't flash. Clicking the status item opens Output Console
- Output Console status item reflects active log count and updates on Clear as well
- Output Console can be launched from View menu in addition to status bar
- Logs show the source of the message (Notebook X, Console Y) and time it was received
- Fixes #1254

### CODE REVIEW

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/frontent-to-kernel-callback/1666/7

**Comment by jasongrout** [comment]
I've been exploring another approach to this, essentially letting a console attached to the notebook's kernel act as a log. The idea is to take care of the specific usecase of comm messages from a notebook having their output/errors silently disappear.

Here are a few things I found:

1. A console doesn't have a `jupyter.widget` target for comms, so if you have a console open and create a widget, the widget is created in the notebook, but the console's kernel connection complains on console.error that the comm target is not registered, and then sends a comm_close message back to the kernel. This means that the kernel's comm will close, orphaning the notebook's comm. The end result is that when you have a console open, new widgets won't work (they are immediately closed, but the notebook frontend doesn't know it). This continues to happen even when the console is closed (since it keeps its kernel connection around?), but there is no clean way to close the kernel connection since the running tab just shows the single session name, the notebook name. So new widgets end up closing immediately and breaking the frontend widgets until you switch kernels or refresh jlab with the console closed. You can see this is all happening if there are lots of errors in the server log saying a comm id is not found (for every thing the notebook tries to sync back to the kernel).
2. Even if you open a console after a widget is created, to avoid the above problem, and enable the "show all kernel activity", you still won't see stdout created by a comm message, since new console cells are only created on execute_input messages: https://github.com/jupyterlab/jupyterlab/blob/97bc2ca68a8f66f3158fb50853218f037b0ecb25/packages/console/src/foreign.ts#L104 . If an output comes in response to a comm message, a new cell is never created, so the output is not printed: https://github.com/jupyterlab/jupyterlab/blob/97bc2ca68a8f66f3158fb50853218f037b0ecb25/packages/console/src/foreign.ts#L115-L118
3. A console does not remember its "Show all kernel activity" status when being restored. This means that if you refresh a page, any console reverts back to the default showing only its own activity. This is confusing.

**Comment by jasongrout** [comment]
Thanks for pushing this forward. It looks really nice. For the record, your approach here was also interesting: mbektasbbg/ipywidgets@fb27ac9

I think best would be between this approach and the ipywidgets one you experimented with. This approach is nice in that it collates output, but it also gives a lot of output with no real way to distinguish between unhandled and handled output. In our architecture, that distinction is at the notebook level (i.e., kernel connection level), not the jlab level. The ipywidgets approach you experimented with is also great, but it only solves the case for ipywidgets, and not more general comm-related output.

Can we explore going right down the middle between these two approaches, similar to the ideas from #1254? What I would imagine is a user being able to open see quickly *just* the unhandled output for a specific notebook, and a button or something in the notebook UX that tells them that there is new unhandled output they haven't seen.

@ivanov suggested it may be easy to take what you have here, and make it easy to filter the view for unhandled messages for a specific kernel connection. (In fact, we probably should by default only store unhandled messages, and only store all messages if the user asks, since that will be a *lot* more output).

You might be able to discern unhandled messages at the notebook kernel connection level when the message comes in (i.e., does it have a callback registered or not?). Or you might be able to do the same trick you did with ipywidgets, where a kernel connection has a default iopub callback, which puts the output in this log widget and signals to the user.

**Comment by jasongrout** [comment]
Looking at this a bit more, I think it's going to be a bit tricky figuring out when a message is unhandled. Essentially, when you send a message, a future is created. Replies to the message are handled by that future: https://github.com/jupyterlab/jupyterlab/blob/de8fa0f57ef24a00e9b9811502146c302d188eb6/packages/services/src/kernel/default.ts#L1248

In our case, we *do* have a future, since we did send a comm message. So in the future, it looks for an iopub handler and hands over the message if it finds one: https://github.com/jupyterlab/jupyterlab/blob/e13121a68727ec9b021e094b2f86148d746a6d9e/packages/services/src/kernel/future.ts#L245

So here is the kicker - widgets *always* defines an iopub handler: https://github.com/jupyter-widgets/ipywidgets/blob/6e5ad194e70a2de8edd30e2757618c0c9e929b4d/packages/base/src/services-shim.ts#L269-L286

So from the perspective of the kernel and the notebook, output messages were handled just fine. If we were really after unhandled outputs, we'd have to have a way for the future iopub handler to say that the message was unhandled, and then the future could go back up and tell the kernel that the message was unhandled, and then the kernel could do something about the message (probably emit it, and then something is listening to the kernel for unhandled iopub messages - I mean, we already have an unhandled message signal at https://github.com/jupyterlab/jupyterlab/blob/de8fa0f57ef24a00e9b9811502146c302d188eb6/packages/services/src/kernel/default.ts#L1257 that is for messages from us that don't have a future - perhaps we just emit the messages there if the future handler says the message wasn't handled).

Thoughts?

**Comment by jasongrout** [comment]
Also, FYI, files in `staging` are updated on release by the release script, and should not be updated manually.

**Comment by mbektasbbg** [comment]
@jasongrout I have a version now that shows the notification on Notebook's toolbar and it filters messages per notebook.

OutputConsoleToolbar

In the capture above, all messages are intercepted. As the next step I can make it catch only unhandled widget messages as in https://github.com/mbektasbbg/ipywidgets/commit/fb27ac9 . We can add a check box to Output Console to catch all messages optionally. Would that match what you had in mind?

**Comment by jasongrout** [comment]
Yes, thanks, this is getting much closer. Do you have thoughts about how to determine which messages are unhandled?

**Comment by mbektasbbg** [comment]
The only way I was able to intercept lost widget messages was as in https://github.com/mbektasbbg/ipywidgets/commit/fb27ac9 , by overriding the default `callbacks`. Do you know a better way?

**Comment by jasongrout** [comment]
My comments above talk about a solution that involves changing how message handling routines can return whether the message was handled or not.

**Comment by jasongrout** [comment]
(my solution above still requires changes in ipywidgets, but the changes are more generic, i.e., returning that we didn't handle a particular message and letting the system handle it, rather than trying to put the message someplace that may not exist in a different spin of jupyterlab, or if widgets are outside the notebook, etc.).

**Comment by mbektasbbg** [comment]
How about if we change all message handler methods/promises to return a status code enum, instead of void? Then we should be able to distinguish unhandled messages.

**Comment by jasongrout** [comment]
> How about if we change all message handler methods/promises to return a status code enum, instead of void?

That's essentially what I had in mind above (except I was thinking true/false, signifying whether the message was actually handled or not). Though I think we need to be more specific about what "all" is in this context. I was imagining just the handlers relevant here, as detailed in tracing through the message handling in https://github.com/jupyterlab/jupyterlab/pull/6833#issuecomment-517551628

**Comment by mbektasbbg** [comment]
I was thinking 'all' needs to cover the calls from this chain https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/src/kernel/default.ts#L1243

Also, there are messages being emitted using signals, like in https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/src/kernel/default.ts#L1252 , do we need to consider handling of messages in those broadcasts?

**Comment by jasongrout** [comment]
> do we need to consider handling of messages in those broadcasts?

No

> I was thinking 'all' needs to cover the calls from this chain https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/src/kernel/default.ts#L1243

So the top-level message handling would return whether the message was handled overall?

**Comment by jasongrout** [comment]
Thinking out loud again: really, at heart here, we need a way to signal that the user hasn't seen a particular output yet. Question is if it's the system's job to get the user's attention, or if is the handler's job. Or put another way, if I tell the system to give me messages and I'll take care of them, can I change my mind and hand them back to the system and say, "just kidding, I won't handle this, you take care of it"?

The other approach is to have a logging system, or a place I *can* give messages that I can't directly handle, or that I can put other errors that I'd like the user to see. For example, as the jlab widget manager, I could log unhandled iopub messages to the js console. That's a logging system that I know will be available. Perhaps there should be a logging console available in jlab, like you are designing here and @ivanov mentioned in conversations, but as a service in jlab for whoever wants to log whatever, rather than tied to specific message handling. And make it the jlab manager's responsibility to log things there. Similarly, any other library that sends a comm message, it is responsible for handling output, including logging that output if it doesn't have a place to display it.

We had a similar discussion with mimerendering - can a mimerenderer claim to handle a certain type of message, but then when handed the message, can it renege on its commitment and hand it back to the system for further processing? We decided for mimerenderers that that wasn't possible - if you say you can handle messages, you better handle them, full stop.

**Comment by mbektasbbg** [comment]
> So the top-level message handling would return whether the message was handled overall?

I guess if the message isn't handled in whole `this._handleMessage(msg)` chain then we can emit unhandledMessage signal but only from a single place like line https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/src/kernel/default.ts#L1243 and not anymore from https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/src/kernel/default.ts#L1286 

Then listeners like Output Console extension can display those unhandled messages.

**Comment by mbektasbbg** [comment]
> We had a similar discussion with mimerendering - can a mimerenderer claim to handle a certain type of message, but then when handed the message, can it renege on its commitment and hand it back to the system for further processing? We decided for mimerenderers that that wasn't possible - if you say you can handle messages, you better handle them, full stop.

I think in the case of ipywidgets, jupyterlab-manager declares, by default it doesn't handle messages if they are not direct output of cell (https://github.com/mbektasbbg/ipywidgets/blob/master/packages/base/src/manager-base.ts#L190) and no other handler exists either.

**Comment by jasongrout** [comment]
> I think in the case of ipywidgets, jupyterlab-manager declares, by default it doesn't handle messages if they are not direct output of cell (https://github.com/mbektasbbg/ipywidgets/blob/master/packages/base/src/manager-base.ts#L190) and no other handler exists either.

But from JLab's point of view, ipywidgets *is* handling the system because it sets an iopub handler on the comm message future: https://github.com/jupyter-widgets/ipywidgets/blob/6ec38c76fb93be0e840e8dfc96495f6e77f533fe/packages/base/src/services-shim.ts#L269-L286

File: packages/services/src/session/default.ts [hunk 1]
```diff
@@ -577,6 +587,28 @@ namespace Private {
...
+    let running = runningSessions.get(settings.baseUrl) || [];
+    let session = find(
+      running,
+      value => value.kernel && value.kernel.clientId === id
+    );
+    if (session) {
+      return Promise.resolve(session.model);
+    }
+
+    return getSessionModel(id, settings).catch(() => {
```
[the remark this answers is not in the document]
Comment (mbektasbbg) [reply packages/services/src/session/default.ts]: you are right, I will throw error if session not found.

File: packages/services/src/session/manager.ts [hunk 2]
```diff
@@ -173,6 +173,17 @@ export class SessionManager implements Session.IManager {
     await this._pollModels.tick;
   }
 
+  get anyMessage(): ISignal<this, Kernel.IAnyMessageArgs> {
```
[the remark this answers is not in the document]
Comment (mbektasbbg) [reply packages/services/src/session/manager.ts]: adding now

File: packages/services/src/session/manager.ts [hunk 3]
```diff
@@ -173,6 +173,17 @@ export class SessionManager implements Session.IManager {
...
+    return this._anyMessage;
+  }
+
+  protected onAnyMessage(
```
[the remark this answers is not in the document]
Comment (mbektasbbg) [reply packages/services/src/session/manager.ts]: will make it private

**Comment by jasongrout** [comment]
@mbektasbbg and I talked on the Jupyter zoom channel about this. We decided based on thinking about things, and various conversations over the last little while, to remove the concept of unhandled messages and instead make the output console a standalone thing that ipywidgets (or whatever else) will send messages to. This is based on a few principles:

1. *The thing with the most relevant information should be the thing to decide what happens.* The concept of an unhandled message is vague, and it isn't clear what should be done at the system level with an unhandled message (the system doesn't know *why* it is unhandled, just that it is supposed to do something, but who knows what the most relevant thing is?). In contrast, the ipywidgets extension, for example, knows that it received an output message, but has no clear place to put the output message, but it still wants the output message displayed to the user. So - ipywidgets itself should display the message to the user in some way. Let's give it a way (i.e., an output console).
2. *Plugins take responsibility for their actions.* This is about simplifying expectations. If something creates a future from a kernel message, it is implicitly promising to handle output from that future. We should empower plugins to take responsibility and fulfill it, rather than encouraging them to give up and hand things back to the system (i.e., unhandled messages). This principle is typically a good idea because of the point above (the plugin has the most relevant information), but is a bit different, in that it is about simplifying the main codebase to not have to deal with cleaning up when plugins don't really have to fulfill the expectations. The data flow and responsibility is clear here - if there is a future, the future handles output, full stop. That simplifies expectations for everyone.

@mbektasbbg said he will also work on making filtering more powerful, with this as a first draft of the UX:

1. The output console is a singleton
2. The output console can be filtered by source name (which is the session name for ipywidgets messages). This filtering can be done programmatically, by a dropdown in the output console UX, and by clicking on a source name in the outputs in the output console. The output console also provides a way to remove the filter in its toolbar or programmatically.
3. The notebook toolbar button will filter the output console to outputs tied to that notebook and open/activate the output console view. Switching notebooks will not automatically filter the output console view.

Did I get everything, @mbektasbbg?

**Comment by mbektasbbg** [comment]
Thanks @jasongrout , yes this covers everything we discussed!

**Comment by mbektasbbg** [comment]
@jasongrout here is the latest version with new filtering options. Note that I am catching all notebook messages only for demo.

OutputConsole3

**Comment by mbektasbbg** [comment]
here is the test code I used :
```
const messageCanBeRendered = (msg: any): boolean => {
    return (
        msg.channel === 'iopub' &&
        [
            'execute_result',
            'display_data',
            'stream',
            'error',
            'update_display_data'
        ].includes(msg.header.msg_type)
    );
};
panel.session.iopubMessage.connect((sender: any, args: any) => {
    if (!messageCanBeRendered(args)) {
        return;
    }
    outputConsoleWidget.outputConsole.logMessage({
        sourceName: panel.session.name,
        sourceIconClassName: 'jp-NotebookIcon',
        msg: args
    });
});

```

added to https://github.com/mbektasbbg/jupyterlab/blob/outputconsole/packages/outputconsole-extension/src/index.tsx#L275

**Comment by afshin** [comment]
Could we add a user setting for the number of log entry items that are stored in memory -- with a default value.

**Comment by jasongrout** [comment]
Also @tgeorgeux requested that we make the notification flash user-configurable (so users can turn it off).

@mbektasbbg - I'm happy to do these things.

**Comment by jasongrout** [comment]
Oh, and they also suggested we hide the button if there are 0 messages for that particular notebook.

**Comment by mbektasbbg** [comment]
@jasongrout sure, please feel free to make any changes as needed.

**Comment by tgeorgeux** [comment]
@mbektasbbg let's use the list_alt icon from material design.
https://material.io/resources/icons/?search=list&icon=list_alt&style=baseline

**Comment by mbektasbbg** [comment]
@jasongrout here is the capture of latest version. 
LogConsole

**Comment by jasongrout** [comment]
@mbektasbbg and I reviewed this in a series of video conference calls over the Jupyter video conference channel. It looks good to me! Thanks!

**Comment by jasongrout** [comment]
Test failures are known and unrelated.

Thanks again @mbektasbbg!

**Comment by jasongrout** [comment]
@meeseeksdev backport to 1.x

**Comment by jasongrout** [comment]
For fun, here is an extension that logs all notebook output:

```typescript
import {
  JupyterFrontEnd,
  JupyterFrontEndPlugin
} from '@jupyterlab/application';
import { ILoggerRegistry } from '@jupyterlab/logconsole';
import { INotebookTracker, NotebookPanel } from '@jupyterlab/notebook';
import { KernelMessage } from '@jupyterlab/services';
import { nbformat } from '@jupyterlab/coreutils';

/**
 * The Log Console extension.
 */
export const logNotebookOutput: JupyterFrontEndPlugin<void> = {
  activate: activateNBOutput,
  id: 'logconsole:nboutput',
  requires: [ILoggerRegistry, INotebookTracker],
  autoStart: true
};

function activateNBOutput(
  app: JupyterFrontEnd,
  loggerRegistry: ILoggerRegistry,
  nbtracker: INotebookTracker
) {
  function registerNB(nb: NotebookPanel) {
    nb.context.session.iopubMessage.connect(
      (_, msg: KernelMessage.IIOPubMessage) => {
        if (
          KernelMessage.isDisplayDataMsg(msg) ||
          KernelMessage.isStreamMsg(msg) ||
          KernelMessage.isErrorMsg(msg) ||
          KernelMessage.isExecuteResultMsg(msg)
        ) {
          const logger = loggerRegistry.getLogger(nb.context.path);
          logger.rendermime = nb.content.rendermime;
          const data: nbformat.IOutput = {
            ...msg.content,
            output_type: msg.header.msg_type
          };
          logger.log({ type: 'output', data });
        }
      }
    );
  }
  nbtracker.forEach(nb => registerNB(nb));
  nbtracker.widgetAdded.connect((_, nb) => registerNB(nb));
}
```

**Pull request merged**